04-onion-routing.md

```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
+
+This raises the question of how Alice knows whether a message was sent
+via a path she supplied.  One method is to remember the final
+`blinding` expected, but requires remembering every such blinding.
+This is the purpose of the `self_id` field: Alice can place a secret
+inside this field, knowing only she can read it, and thus verify the
+path as one she created.  It is variable length so she can also
+add data to identify which path it was.
+
+#### Requirements
+
+The writer:
+- For the non-final nodes' `onionmsg_payload`:
+  - MUST set `enctlv` to a valid `encmsg_tlvs` containing `next_node_id`.
+  - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
+    `ChaChaPoly-1305` encryption using an all-zero nonce).
+  - MAY include `padding`.
+  - MUST NOT set `self_id`.
+- For the final node's `onionmsg_payload`:
+  - if the final node is permitted to reply:
+    - MUST set `reply_path` `blinding` to the initial blinding factor for the `next_node_id`
+    - MUST set `reply_path` `first_node_id` to the unblinded node id of the first node in the reply path.
+    - For every `reply_path` `path`:
+      - MUST set `node_id` to the blinded node id to encrypt the onion hop for.
+      - MUST encrypt `enctlv` as detailed in [Route Blinding][route-blinding].
+        `ChaChaPoly-1305` encryption using an all-zero nonce).
+      - MUST set `enctlv` to a valid `encmsg_tlvs` which meets the requirements
+        of the `onionmsg_payload` when used by the recipient.
+      - MAY use `self_id` to contain a secret so it can recognize use of this `reply_path`.
+  - otherwise:
+    - MUST NOT set `reply_path`.
+
+The reader:
+- if it is not the final node according to the onion encryption:
+  - if `enctlv` is not present, or does not decrypt with the shared secret from the given `blinding` parameter:
+    - MUST drop the message.
+  - if the `enctlv` is not a valid TLV or does not contain `next_node_id`:
+    - MUST drop the message.
+  - if the `enctlv` contains `self_id`:
+    - MUST drop the message.
+  - otherwise:
+    - MUST ignore `padding`, if any.
+    - SHOULD forward the message using `onion_message` to the next peer indicated by `next_node_id`.
+    - if it forwards the message:
+      - if `blinding` is specified in the `enctlv`:
+         - MUST pass that as `blinding` in the `onion_message`
+      - otherwise:
+         - MUST pass `blinding` derived as in [Route Blinding][route-blinding] (i.e. `E(i+1) = H(E(i) || ss(i)) * E(i)`).
+
+- otherwise:
+  - if this was sent via a `reply_path` the reader previously supplied:
+    - if it is not a reply to the onion message sent with that `reply_path`
+      - MUST ignore the onion message.
+  - otherwise:
+    - if this is a reply to an onion message which was sent with a `reply_path`:
+      - MUST respond (or not respond) exactly as if it did not send the initial onion message.
+  - if it wants to send a reply:
+    - MUST create an onion encoding using `reply_path`.
```

Comment:
This needs to be explained with more details.

Reply:
Yes...